build.gradle

```diff
@@ -1,59 +1,172 @@
+    args '-li2c'
+
 }
+
 tasks.register('createJhwbusResourceDir', Sync) {
+    description = 'Create jhwbus resource directory'
     from tasks.named('compileC')
     into project.layout.buildDirectory.dir('jhwbus')
 }
+
 sourceSets.main.resources.srcDirs tasks.named('createJhwbusResourceDir')
 
-repositories {
-    mavenCentral()
+jar {
+    enabled = OS_NAME == 'linux'
 }
 
-dependencies {
-    testImplementation 'org.spockframework:spock-core:2.0-groovy-3.0'
-    testImplementation 'org.codehaus.groovy:groovy-all:3.0.8'
+String image =  project.properties['dockerImage'] ?: "ghcr.io/org-arl/jhwbus:latest"
+String dockerContext = project.properties['dockerContext'] ?: 'colima'
+tasks.register('buildImage'){
+    description = 'Builds a docker image'
+    doLast {
+        exec {
+            commandLine 'docker', 'build', '-t', image, '.'
+        }
+    }
+}
+
+tasks.register('pushImage'){
+    description = 'Pushes the docker image to GitHub Packages'
+    dependsOn 'buildImage'
+    doLast {
+        exec {
+            commandLine 'docker', 'push', image
+        }
+    }
+}
+
+project.ext.container = null
+tasks.register('startContainer') {
+    description = 'Starts a container'
+    doLast {
+        if (project.ext.container instanceof String) return
+
+        // start a container using docker cli and exec
+        println "Starting a new container ($image)... "
+        project.ext.container = execWithOutput('docker','--context', dockerContext , 'run', '-d', '-v', "$rootDir:/home/jhwbus", image, 'sleep', 'infinity')
+    }
+}
+
+tasks.register('stopContainer') {
+    description = 'Stops the container'
+    doLast {
+        if (!(project.ext.container instanceof String)) return
+        // stop the container using docker cli and exec
+        execWithOutput('docker', '--context', dockerContext , 'stop', project.ext.container as String)
+        println "Stopped the container"
+        project.ext.container = null
+    }
+}
+
+tasks.register('containerJar') {
+    description = 'Builds jhwbus in a container'
+    dependsOn tasks.startContainer
+    finalizedBy tasks.stopContainer
+    doLast {
+        def containerId = project.ext.container
+        if (!containerId || !(containerId instanceof String)) {
+            throw new GradleException("No known container!")
+        }
+        // check if the container is running using docker cli
+        def running = execWithOutput('docker', '--context', dockerContext, 'inspect', '-f', '{{.State.Running}}', containerId as String)
+        if (running != 'true') {
+            throw new GradleException("Container is not running!")
+        }
+        // exec the build command in the container
+        println "Running build in a container : ${containerId.take(12)}"
+        def res = exec {
+            commandLine 'docker', '--context', dockerContext , 'exec', containerId, '/bin/bash', '-c', "./gradlew -Dorg.gradle.welcome=never jar"
+            ignoreExitValue true
```

Comment:
Why are we ignoring the exit values (both here and in `execWithOutput`)? I assume this means that it is possible to get a "Build succeeded" when actually the build failed?

Reply:
https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Exec.html#org.gradle.api.tasks.Exec:ignoreExitValue

Gradle's ExecSpec has a bit confusing name. If `ignoreExitValue` is `false`, the command failing throws an Exception. If `ignoreExitValue` is `true`, it just returns the result, which we then check and deal with.

I generally prefer dealing with return codes than Exceptions.

Reply:
Right, I missed that we check the return code just after that. Nvm, then 🙈